修改 Thread.isMainThread

the-yuyut opened the pull request:
根據 這篇文章: [GCD's Main Queue vs. Main Thread](http://blog.benjamin-encz.de/post/main-queue-vs-main-thread/?fbclid=IwAR2W9FHQUIntKF2WwwBWR4yet4k3StuBO8E-SqSTtIQ7fJYlgyXQ9rJgXoE)

我研究了一下這麼在 Swift 5 實作他。

更新，我後來改用 label 來作為 Equable 的處理，透過 Objective-C的 `__dispatch_queue_get_label` 拿到 Current Queue 的 label。

the-yuyut committed:
rewrite MainThreadDecorator by http://blog.benjamin-encz.de/post/main-queue-vs-main-thread/

the-yuyut committed:
apply test on MainThreadsTests

the-yuyut commented:
抱歉，寫完測試才發現根本不需要set，等等來改

dsxsxsxs commented on Practical Dependency InversionTests/MainThreadDecoratorTests.swift:
```diff
@@ -0,0 +1,59 @@
+}
+
+
+func checkQueue(queue: DispatchQueue, result: inout Bool?){
```
[Q]
why use inout? Just return it then.

the-yuyut replied:
No idea why I prefer using `inout`, but I don't know why not.

the-yuyut replied:
Yap, it turns out you are right! Returning the value is pretty neat!

dsxsxsxs commented on Practical Dependency InversionTests/MainThreadDecoratorTests.swift:
```diff
@@ -0,0 +1,59 @@
+        let testQueue = DispatchQueue(label: #function)
+        let await = XCTestExpectation(description: #function)
+        var result:Bool?
+        testQueue.async {
```
[Q]
Is this necessary?(dispatch to another queue) What's the purpose?

the-yuyut replied:
The main purpose of this test is to test some statement should run on the main queue. To do that, we need to make sure that the current queue must not main queue, that just reminds me that I did not test for it. I will do some modifications.

the-yuyut replied:
Added the assertion before redirecting to the main queue, make sure there are different.

dsxsxsxs replied:
Ah! Now that makes perfect sense. Looks great!

the-yuyut replied:
Thanks a lot!

the-yuyut committed:
point out no matter what, the MainTHreadDecorator will always on the same queue of caller, because it will be call first.

the-yuyut committed:
rewrite to handle dispatch on loader callback

the-yuyut commented:
發現原本實際上的 Call stack 在回呼之前，所以改成在 callback 之後才會做 queue 的重新導向。

dsxsxsxs commented on Practical Dependency Inversion/ItemsUIComposer.swift:
```diff
@@ -35,12 +35,14 @@ class MainThreadDecorator: ItemsViewControllerSpec {
-            return
+        loader.items {
+            result in
+            DispatchQueue.main.async {
```
So, This means that there is no need to identify which the current queue is, always dispatch to main queue would be good. Am I right?
Pretty sensible way, I think.👍

OTOH
So I guess that this means that you can get rid of both [this](https://github.com/rollr76518/Practical-Dependency-Inversion/pull/1/commits/8b7983ef600c218655afdeb73145abc57ba83373#diff-404da706489f4f16c0022da0e6c41898R24-R26) and [this](https://github.com/rollr76518/Practical-Dependency-Inversion/pull/1/commits/8b7983ef600c218655afdeb73145abc57ba83373#diff-404da706489f4f16c0022da0e6c41898R30-R34)?
The test code might also become kind of meaningless, perhaps?

the-yuyut replied:
Sorry, can't really understand what's that means.
> So I guess that this means that you can get rid of both this and this?

And the test code for `isMainQueue` is just to make sure which will always give the right result.

the-yuyut replied:
> This means that there is no need to identify which the current queue is, always dispatch to main queue would be good. 

It's still questionable, but it seems harmless by dispatch without checking.

dsxsxsxs replied:
> Sorry, can't really understand what's that means.
> 
> > So I guess that this means that you can get rid of both this and this?
> 
> And the test code for `isMainQueue` is just to make sure which will always give the right result.

Just tried to pick line 24 ~ 26, line 30 ~ 34 up but it ain't work at all.😂
U don't need to specify the main queue since those main queue checking things in production code is no longer needed, do u?

the-yuyut replied:
Absolutely, thanks for point out. I will remove those statements.

the-yuyut committed:
remove setSpecific

Because we dispatch directly to `main queue` without checking, set specific became pointless.

rollr76518 commented:
麻煩幫我把 file sorting 一下。(無聊的龜毛)

the-yuyut commented:
@rollr76518 
> 麻煩幫我把 file sorting 一下。(無聊的龜毛)

我看起來是已經排序好了，不太確定你是不是說這個部分。

Practical Dependency Inversion.xcodeproj/project.pbxproj line 249~250
```
files = (
	A7F406A624BDB931005E381D /* MainThreadDecoratorTests.swift in Sources */,
	7EDDA12724B16093003393C1 /* ItemsViewControllerTests.swift in Sources */,					
);
```

dsxsxsxs reviewed (approved):
LGTM

allenlinli commented:
感謝 @ytyubox 分享，學到一課。以後遇到這種情況在 async 就不管是否已經在 mainThread 上而直接做了。

the-yuyut commented:
@rollr76518 如果還有要改的部分，可以用 GitHub 的 Change request 來討論喔
reference： [Incorporating feedback in your pull request - GitHub Docs](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request)

rollr76518 commented:
非常感謝！
